### PR TITLE
[Reopen] Read args from file, merge args and pass to the cmd. 

### LINKED ISSUE #191: Add support for reading command-line options from file(s)
The users command-line app has a top-level command, whose name is `<name>`.

For example, if the user was creating a word count CLI, then their top-level command name might be `wc`.

ZIO CLI should examine the following locations to check for the existence of `.<name>` files:

1. The current working directory.
2. All parents of the current working directory.
3. The user's home directory.

So, in the `wc` example, if run from `/c/Users/john` directory, then ZIO CLI would look in the following locations for the existence of `.wc` files:

1. `/c/Users/john`
2. `/c/Users/
3. `/c/`
4. `~` (user home, probably `/c/Users/john` on Windows)

Then, ZIO CLI should concatenate these options, with user home having the lowest priority, and the current working directory having the highest priority (so, for example, you can override options in higher priority locations).

Then, ZIO CLI should utilize those command-line options when the program is run, and not require those options be specified. However, if any option is specified on the command-line, then it must override even the highest priority setting specified in the settings files.

An example, let's say `wc` has an option `--line-ending`. Then in the `.wc` file, there might exist:

```
--line-ending=\n
```

Then this would be the default line ending. The user would not have to specify it on the command-line. But if the user did specify it on the command-line, it would override the one in the file.

This new functionality should be documented in the help docs for the CLI. Moreover, all of the settings that are picked up from any of the files need to be printed out, so the user can understand exactly where they are coming from.

### PR DESCRIPTION
**Pull request by vivasvan1**
/claim #191 

closes #191 

Hi @jdegoes,

Hi have attempted to resolve your [review comments on the previous PR](https://github.com/zio/zio-cli/pull/278#:~:text=This%20creates%20a%20leak%20because%20the%20source%20is%20never%20closed.%20Also%2C%20it%27s%20an%20inefficient%20way%20to%20read%20a%20file.%20Better%20to%20use%20something%20like%20ZIO.readFile.)

Please let me know if more changes are required.

Working Screenshots:
image

image

---
References: 
1) Old PR: https://github.com/zio/zio-cli/pull/278

P.S. : I dont have permission to reopen the previous PR so created a fresh one.

### CODE REVIEW

**Comment by vivasvan1** [comment]
Also could i please get provide some guidance in fixing CI/CD error. I am new to this ecosystem and unclear on what exactly those mean.

**Comment by jdegoes** [comment]
```
[error] Modules were resolved with conflicting cross-version suffixes in ProjectRef(uri("file:/home/runner/work/zio-cli/zio-cli/"), "zioCliJVM"):
[error]    org.scala-lang.modules:scala-collection-compat _3, _2.13
[error] java.lang.RuntimeException: Conflicting cross-version suffixes in: org.scala-lang.modules:scala-collection-compat
```

This error is related to Scala JS, to use of a library scala-collection-compat. Probably version numbers for the Scala compiler(s) need updating to be consistent with each other. I would look in the build file.

**Review by jdegoes — CHANGES_REQUESTED** [review]
Looks good! Please just add a test to ensure the behavior is working correctly, and th en we can merge!

**Comment by vivasvan1** [comment]
> Looks good! Please just add a test to ensure the behavior is working correctly, and th en we can merge!

@jdegoes 
Update:
Hi there,

First, thank you for your review.

I have commited fix for build errors in JVM and Native.

However, errors are now only in scalaJS.
Error are only of this type:
image

I believe it is JS doesnt support `java.io`
If you know about this could you guide me a little?

Also, I'm working on writing test. should be done soon. 

Thank you again :)

**Comment by vivasvan1** [comment]
@jdegoes 
Hi there, hope this message finds you well.
Reminder: I have fixed the issues you mentioned.
Could you please let me know how to proceed?
Thanks :)

**Comment by vivasvan1** [comment]
@pablf Hi there,

I have added the test case you asked for. Where the functionality is being tested on the full wc app. 
Let me know if any futher changes are required.

Thanks,
Vivasvan

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        // Check if the func checkAndGetOptionsFilePaths can
+        paths <- configFileOps.findPathsOfCliConfigFiles(command)
+
+        _ <- cleanUpSampleConfigFiles(cwd: Path, homeDir: Path, command)
+
+      } yield assert(paths)(hasSameElements(List(homeDir.toString(), cwd.toString())))
+    }
+  )
+
+  def createLineCountTestFile(
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:75]: This should return
```scala
ZIO[Scope, IOException, Unit]
```
and clean itself up

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    cwd: Path,
+    file_name: String = "sample_file",
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:77]: `fileName`

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+      Files.delete(Paths.get(cwd.toString(), s"$file_name"));
+      ()
+    }.refineToOrDie[IOException]
+
+  def createSampleConfigFiles(
+    cwd: Path,
+    homeDir: Path,
+    command: String = "testApp",
+    content_home: String = "home=true",
+    content_cwd: String = "dir=true\nhome=false"
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:96]: audit entire file please

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  ): IO[IOException, Unit] =
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:97]: same thing as above, writing to file should return zio `Scope`, and clean itself up, doing so for a single file, and be called twice

File: zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.IOException
+import zio._
+import java.nio.file.{Files, Path, Paths}
+import scala.io.Source
+
+object ConfigFileArgsPlatformSpecific extends ConfigFilePlatformSpecific {
+  def findPathsOfCliConfigFiles(topLevelCommand: String): Task[List[String]] = {
+    val filename   = s".$topLevelCommand"
+    val cwd        = java.lang.System.getProperty("user.dir")
+    val homeDirOpt = java.lang.System.getProperty("user.home")
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala:12]: See: `zio.System`

File: zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                 .foreach(filePaths) { filePath =>
+                   ZIO.acquireReleaseWith(
+                     ZIO.attempt(
+                       Source.fromFile(Paths.get(filePath, "." + topLevelCommand).toFile)
+                     )
+                   )(source => ZIO.attempt(source.close()).orDie) { source =>
+                     ZIO.attempt(source.getLines().toList.filter(_.trim.nonEmpty))
+                   }
+                 }
+                 .map(_.flatten)
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala:45]: `zio.ZIO.readFile`

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        _ <- cleanupLineCountTestFile(cwd, "sample_file")
+        _ <- cleanUpSampleConfigFiles(cwd: Path, homeDir: Path, command)
+      } yield assert(output.last)(equalTo(correctOutput))
+    },
+    test("should load options from files and merge them appropriatly") {
+      for {
+        // Create Sample config files
+        cwd     <- ZIO.succeed(Paths.get(java.lang.System.getProperty("user.dir")))
+        homeDir <- ZIO.succeed(Paths.get(java.lang.System.getProperty("user.home")))
+        command <- ZIO.succeed("testApp")
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:48]:
```scala
command = "testApp"
```

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+  val configFileOps: ConfigFilePlatformSpecific = ConfigFileArgsPlatformSpecific
+
+  def spec = suite("FileBasedArgs")(
+    test("Full CLI App output test") {
+      for {
+        // Create Sample config files
+        cwd     <- ZIO.succeed(Paths.get(java.lang.System.getProperty("user.dir")))
+        homeDir <- ZIO.succeed(Paths.get(java.lang.System.getProperty("user.home")))
+        command <- ZIO.succeed("wc")
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:29]:
```scala
command = "wc"
```

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+        _ <- createLineCountTestFile(cwd, "sample_file")
+        _ <- createSampleConfigFiles(cwd, homeDir, command, content_home = "-w", content_cwd = "-l")
+
+        _      <- cliApp.run(args = List("sample_file")).either
+        output <- TestConsole.output
+
+        correctOutput <- ZIO.succeed("        3         4        20  sample_file\n")
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:37]: `=`

File: zio-cli/shared/src/main/scala/zio/cli/CliApp.scala [hunk 10]
```diff
@@ -128,19 +128,30 @@ object CliApp {
...
-              executeBuiltIn(x).catchSome { case err @ CliError.Parsing(e) =>
-                printDocs(e.error) *> ZIO.fail(err)
-              }
+      // Reading args from config files and combining with provided args
+      val combinedArgs: ZIO[R, CliError[E], List[String]] =
+        ConfigFileArgsPlatformSpecific
+          .loadOptionsFromConfigFiles(self.command.names.head)
+          .flatMap { configArgs =>
+            ZIO.succeed(configArgs ++ args)
           }
```
Comment (Kalin-Rudnicki) [inline zio-cli/shared/src/main/scala/zio/cli/CliApp.scala:137]: `.map`

File: zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala [hunk 11]
```diff
@@ -0,0 +1,48 @@
+package zio.cli
+
+import java.io.IOException
+import zio._
+import java.nio.file.{Files, Path, Paths}
+import scala.io.Source
+
+object ConfigFileArgsPlatformSpecific extends ConfigFilePlatformSpecific {
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala:8]: This seems to be copy-pasted, move into `./zio-cli/jvm-native`

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    }
+  )
+
+  def createLineCountTestFile(
+    cwd: Path,
+    file_name: String = "sample_file",
+    content: String = "asdf\nqweqwer\njdsafn"
+  ): IO[IOException, Unit] =
+    ZIO.attempt {
+      Files.write(Paths.get(cwd.toString(), s"$file_name"), java.util.Arrays.asList(content));
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:81]: `content.getBytes`

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala (hunk 3 above)
Comment (Kalin-Rudnicki) [reply zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:96]: Also, is it possible to remove the default arguments?  
I find reading through the tests very hard to follow, and easily determine what output I would expect.

Also, I imagine that there should be multiple tests asserting expected behavior for different permutations of hierarchy overrides.

File: zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala (hunk 12 above)
Comment (Kalin-Rudnicki) [reply zio-cli/jvm/src/test/scala/zio/cli/FileBasedArgs.scala:81]: `s"$file_name"` = `fileName`

File: zio-cli/native/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    val filename   = s".$topLevelCommand"
+    val cwd        = java.lang.System.getProperty("user.dir")
+    val homeDirOpt = java.lang.System.getProperty("user.home")
+
+    def parentPaths(path: String): List[String] = {
+      val parts = path.split(java.io.File.separatorChar).filterNot(_.isEmpty)
+      (0 to parts.length)
+        .map(i => s"${java.io.File.separatorChar}${parts.take(i).mkString(java.io.File.separator)}")
+        .toList
+    }
```
Comment (Kalin-Rudnicki) [inline zio-cli/native/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala:19]: Does this work for `Windows`/`Linux`/`Mac`?  
I have a sneaking suspicion it might not work correctly for `/home/kalin/dev/a/b/c/.cmd`  
Might be safer to implement via `Path#getParent`?

File: zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    val filename   = s".$topLevelCommand"
+    val cwd        = java.lang.System.getProperty("user.dir")
+    val homeDirOpt = java.lang.System.getProperty("user.home")
+
+    def parentPaths(path: String): List[String] = {
+      val parts = path.split(java.io.File.separatorChar).filterNot(_.isEmpty)
+      (0 to parts.length)
+        .map(i => s"${java.io.File.separatorChar}${parts.take(i).mkString(java.io.File.separator)}")
+        .toList
+    }
```
Comment (Kalin-Rudnicki) [inline zio-cli/jvm/src/main/scala/zio/cli/ConfigFileArgsPlatformSpecific.scala:19]: Does this work for `Windows`/`Linux`/`Mac`?  
I have a sneaking suspicion it might not work correctly for `/home/kalin/dev/a/b/c/.cmd`  
Might be safer to implement via `Path#getParent`?

File: zio-cli/shared/src/main/scala/zio/cli/ConfigFileArgs.scala [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package zio.cli
+
+import zio._
+import java.io.IOException
+
+trait ConfigFilePlatformSpecific {
+  def findPathsOfCliConfigFiles(topLevelCommand: String): Task[List[String]]
+  def loadOptionsFromConfigFiles(topLevelCommand: String): ZIO[Any, IOException, List[String]]
+}
```
Comment (Kalin-Rudnicki) [inline zio-cli/shared/src/main/scala/zio/cli/ConfigFileArgs.scala:9]: Personal opinion:  
I think it would be nicer if this passed via something `~like`
```scala
trait ConfigFileArgs
object ConfigFileArgs {
  val default: ConfigFileArgs // platform specific
}
```
```scala
sealed trait CliApp[-R, +E, +A] { self =>
  def runWithFileArgs(args: List[String]): ZIO[R & ConfigFileArgs, CliError[E], Option[A]]
  final def run(args: List[String]): ZIO[R, CliError[E], Option[A]] = 
    runWithFileArgs(args).provideSomeLayer(ZLayer.succeed(ConfigFileArgs.default))
```

**Comment by vivasvan1** [comment]
@Kalin-Rudnicki 
Wow, thank you so much for all the comments, I really appreciate your help 🙏.

I will look into this asap. This seems really helpful.
